### ISSUE TITLE
mergeProps with queryProps/mutationProps

### ISSUE THREAD

**Issue by deoqc** [apollostack/react-apollo#62]
Currently the `mergeProps` function doesn't allow for `queryProps` or `mutationProps`.

Will it be different? I find useful to keep the logic in this higher order components. Usually only use `map*ToProps` but the `mergeProps` is a must when needed.

**Comment by stubailo** [comment]
Can you give an example of a use case for this? Just want to understand it better.

**Comment by deoqc** [comment]
It is never a need but sometimes a must. 

tl;dr: It helps to keep code more organized (*presentational* vs *container* component), and as a side effect may have good optimizations with regard to re-render.

---

I have 2 use cases:

###### Convenient place to make calculations
I have a `mapQueryToProps` that get informations of products and have shopping cart data only on client. If I want to calculate the total (combining info from these both places), I would like to do it in the `mergeProps` and pass to my *presentational* component only the precise data to render.

Of course, I can do this calculation in the (previously *presentational*) component, or even have other higher order *container* component only for this. But I would prefer to do it in `mergeProps` (or something similar).

###### Convenient place to avoid re-render
If I pass the whole query data to the component, any minor change will trigger a re-render of the component. Sometimes I use a calculated value that may not change even if the data as whole do (as in the first use case).

Again, I could use `shouldComponentUpdate` or have other higher order *container* component do this. But `mergeProps` would be much better.

**Comment by jbaxleyiii** [comment]
@deoqc so are you saying having access to the props from `mapQueriesToProps` and `mapMutatationsToProps` in `mergeProps`?

Currently react-apollo does `redux-connect [mapStateToProps, mapDisptachToProps` => `apollo-connect [mapQueriesToProps, mapMutationsToProps]`. I could see how switching them could be helpful since apollo-connect has the entire redux state, but `mapStateToProps` doesn't get the new props from your GraphQL needs.

I am a little concerned about over rendering since `react-redux` does a lot to try and not over-render child components. 

So if I understand you correctly I'm game and will try and see how we can keep it performant!

**Comment by deoqc** [comment]
That's right, it's just having access to mapQueriesToProps and
mapMutationsToProps in mergeProps.
Em sáb, 28 de mai de 2016 às 18:55, James Baxley <notifications@github.com>
escreveu:

**Comment by stubailo** [comment]
I just heard about this at React Europe: https://github.com/acdlite/recompose

Could be a good utility to suggest for people who want to transform their query results before passing them down!

**Comment by jbaxleyiii** [comment]
#29 will abstract out redux-connect, making this possible to compose `connect(apollo(component))` or `apollo(connect(component))`

**Comment by deoqc** [comment]
Hey @jbaxleyiii, what would be the status of this?

I'm currently using [recompose](https://github.com/acdlite/recompose) as suggested by @stubailo, but I used to only need `mergeProps` with [reselect](https://github.com/reactjs/reselect).

**Comment by jbaxleyiii** [comment]
@deoqc the new api #120 separates connect and apollo. So you can `connect(apollo(component))` or `apollo(connect(component))`. It should be released today!

**Comment by deoqc** [comment]
Awesome!
